driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -71,100 +70,92 @@
         this.limit = limit;
         this.batchSize = batchSize;
         this.decoder = decoder;
-        this.cursor = firstBatch.getCursor();
-        if (this.cursor != null) {
-            notNull("connectionSource", connectionSource);
-            notNull("connection", connection);
-        }
-        if (connectionSource != null) {
-            this.connectionSource = connectionSource.retain();
-        } else {
-            this.connectionSource = null;
-        }
+        this.cursor = new AtomicReference<ServerCursor>(firstBatch.getCursor());
+        this.connectionSource = notNull("connectionSource", connectionSource);
         this.count += firstBatch.getResults().size();
-        if (limitReached()) {
-             killCursor(connection);
+
+        connectionSource.retain();
+        if (firstBatch.getCursor() != null && limitReached()) {
+            killCursor(connection);
```

Comment:
There's a bug here, and I see now that this is an untested code path at least in the functional tests.

The problem is that killCursor now releases the connection, but since this connection is not retained here the refcount is going to drop below 0 once the caller releases it.  So there has to be an extra retain somewhere, either here or in killCursor.   Note that the killCursor method currently retains the connection, but that retain is no longer there.